### ISSUE TITLE
The last versions of Nginx Unit fails in the TechEmpower benchmark

### ISSUE THREAD

**Issue by joanhey** [nginx/unit#794]
Hi, 
I added Nginx and Nginx Unit to the TechEmpower benchmark.
But in the last versions it's failing with PHP:
https://tfb-status.techempower.com/unzip/results.2022-11-25-12-16-16-174.zip/results/20221119184855/php-unit/run/php-unit.log

Old config without problems, and faster than Nginx:
https://github.com/TechEmpower/FrameworkBenchmarks/blob/R20/frameworks/PHP/php/php-unit.dockerfile
https://github.com/TechEmpower/FrameworkBenchmarks/blob/R20/frameworks/PHP/php/deploy/nginx-unit.json

Actual config with problems:
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/PHP/php/php-unit.dockerfile
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/PHP/php/deploy/nginx-unit.json

The last versions fail a lot.

Please, could anybody review and help with this problem.
Thank you.

PD: Welcome help with the Nginx config too, only with the plaintext test.
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/C/nginx/nginx.conf

Be careful, as the bench changed a lot after the spectre/meldown patches.
https://github.com/TechEmpower/FrameworkBenchmarks/issues/7321

If you need I'll send you the benchs results. And ask me for any doubt.

**Comment by ac000** [comment]
Looking at the log it seems

```
php-unit: /usr/local/bin/docker-entrypoint.sh: Applying configuration /docker-entrypoint.d/nginx-unit.json
php-unit: 2022/11/23 16:50:03 [info] 21#21 "benchmark" prototype started
```

Unit was started with some initial config.

```
php-unit: 2022/11/23 16:50:03 [info] 105#105 "benchmark" application started
php-unit: /usr/local/bin/docker-entrypoint.sh: OK: HTTP response status code is '200'
php-unit: {
php-unit: 	"success": "Reconfiguration done."
php-unit: }
php-unit: /usr/local/bin/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/...
php-unit: /usr/local/bin/docker-entrypoint.sh: Stopping Unit daemon after initial configuration...
php-unit: /usr/local/bin/docker-entrypoint.sh: Waiting for control socket to be removed...
php-unit: 2022/11/23 16:50:03 [alert] 21#21 sendmsg(20, -1, -1, 1) failed (104: Connection reset by peer)
```

After it was started it was reconfigured, then shutdown. All that happened within the space of a second.

From your point of view, what was the actual sequence of events?

**Comment by joanhey** [comment]
The strange thing is  that in local work.
Show all that verbose output, but work.

Finish here and start to verify the tests:
![image](https://user-images.githubusercontent.com/249085/204663307-600c592b-1f40-4817-a349-24881cda11e6.png)

When finish the tests:
![image](https://user-images.githubusercontent.com/249085/204663407-6ad6e310-c388-45cb-9882-4f145352a3d6.png)

Before show the summary:
![image](https://user-images.githubusercontent.com/249085/204663498-ee3fe6a3-5b15-4a7a-8866-39fef8c30d0e.png)

But not in github actions or the bench server.

Github actions log:
https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/3567839471/jobs/5996024544#step:9:4375

**Comment by joanhey** [comment]
It's failing from the last 2 versions of Nginx Unit.

**Comment by joanhey** [comment]
When I send the PR for Nginx Unit 1.28 was passing the github actions tests.
But not in the bench server, and later again failed in github actions.

https://github.com/TechEmpower/FrameworkBenchmarks/pull/7691

Github actions log working:
https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/3447671893/jobs/5753927577#step:9:2154

**Comment by tippexs** [comment]
As discussed on the community slack https://nginxcommunity.slack.com/archives/C02SS9UB85C/p1670008983875989?thread_ts=1669810946.221959&cid=C02SS9UB85C

This issue is related to #728 and friends. A manual patch of the `docker-entrypoint.sh` file seems to fix the issue. 

Linking the external issue as well. 

https://github.com/TechEmpower/FrameworkBenchmarks/pull/7757#pullrequestreview-1203015505

As said, we will ship a fix for the entrypoint script with the next Unit release 1.29.

**Comment by joanhey** [comment]
If work in the next run. I'll close the isse.

**Comment by tippexs** [comment]
I will close it once the fix was merged if that's okay for you

**Comment by joanhey** [comment]
Failed in the last run:

![image](https://user-images.githubusercontent.com/249085/209589703-bf4182b8-fdc2-411d-8752-30dd722424ed.png)

![image](https://user-images.githubusercontent.com/249085/209589725-2973d249-7c23-41b1-8a39-5f306c9a4b4b.png)

When finish the run, I'll send the logs.

**Comment by tippexs** [comment]
Yes, please share the logs. The new images should have the new docker entrypoint script.
In case this is another kind of issue we should
Create a new one.

**Comment by joanhey** [comment]
It's using the patch in v1.28, and looks like the same issue.
https://tfb-status.techempower.com/unzip/results.2022-12-28-17-59-18-855.zip/results/20221222172047/php-unit/build/php-unit.log

https://tfb-status.techempower.com/unzip/results.2022-12-28-17-59-18-855.zip/results/20221222172047/php-unit/run/php-unit.log

I'll update to v1.29

**Comment by joanhey** [comment]
Failed in the last run, using v1.29.
https://github.com/TechEmpower/FrameworkBenchmarks/pull/7839

Sunit (scala) from @lolgab, also fail with v1.8 and the last commit with v1.29.
https://github.com/TechEmpower/FrameworkBenchmarks/pull/7815
Sunit log with v1.29:
https://tfb-status.techempower.com/unzip/results.2023-01-04-02-03-49-510.zip/snunit


The only framework using v1.28 and working is Fastapi (python):
https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Python/fastapi
Log:
https://tfb-status.techempower.com/unzip/results.2023-01-04-02-03-49-510.zip/results/20221229020235/fastapi-nginx-unit

**Comment by joanhey** [comment]
It's the same issue. Not necessary to open a new one.

**Comment by joanhey** [comment]
Update, Fastapi (python) fail with v1.29 in the tests.

https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/3745949871/jobs/6360874318#step:9:1796

**Comment by tippexs** [comment]
> It's using the patch in v1.28, and looks like the same issue. https://tfb-status.techempower.com/unzip/results.2022-12-28-17-59-18-855.zip/results/20221222172047/php-unit/build/php-unit.log
> 
> https://tfb-status.techempower.com/unzip/results.2022-12-28-17-59-18-855.zip/results/20221222172047/php-unit/run/php-unit.log
> 
> I'll update to v1.29

This issue is different from the other once. It is not related to the control socket. Will have a look on it.

**Comment by travisbell** [comment]
> Update, Fastapi (python) fail with v1.29 in the tests.
> 
> https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/3745949871/jobs/6360874318#step:9:1796

This log has these lines in it:

> fastapi-nginx-unit-orjson: 2022/12/21 02:57:25 [alert] 0#25 [unit] Python failed to call 'asyncio.get_event_loop'
> Notice: fastapi-nginx-unit-orjson: 2022/12/21 02:57:25 [notice] 24#24 app process 25 exited with code 1

Which seems like it's the same issue as outlined in #815 perhaps?

**Comment by tippexs** [comment]
Yes indeed! This is exactly this issue. We debugged the Python code last night. 1.29 is working with Python 3.10.7. anything newer than this will currently fail. We are working on a fix and I will let you know once available. Thanks for the links.

**Comment by micaelmalta** [comment]
> Yes indeed! This is exactly this issue. We debugged the Python code last night. 1.29 is working with Python 3.10.8. anything newer than this will currently fail. We are working on a fix and I will let you know once available. Thanks for the links.

Any progress in this issue?

**Comment by ac000** [comment]
@micaelmalta (and anyone else) feel free to try the patch here https://github.com/nginx/unit/issues/815#issuecomment-1396132323

**Comment by joanhey** [comment]
V1.29.1 still fail in the benchmark tests, almost with PHP.
When finish the run, I'll send the logs.

**Comment by joanhey** [comment]
Logs:
https://tfb-status.techempower.com/unzip/results.2023-03-10-06-06-32-467.zip/php-unit

https://tfb-status.techempower.com/unzip/results.2023-03-10-06-06-32-467.zip/fastapi-nginx-unit


Check build and run dirs

**Comment by tippexs** [comment]
@joanhey Thanks. We will have a look into this. More details on Monday.

**Comment by lolgab** [comment]
My Scala Native benchmark using `libunit` [works](https://www.techempower.com/benchmarks/#section=test&runid=0f03793d-ac4c-4869-b1fa-51fcef26d48f&test=plaintext&hw=cl&p=zik0zj-zik0zj-zik0zj-zik0zj-8vmzun) using NGINX Unit `1.29.1`.

**Comment by joanhey** [comment]
Work but with a lot of errors.

Run logs:
- snunit: https://tfb-status.techempower.com/unzip/results.2023-03-23-04-22-59-528.zip/results/20230317041717/snunit/run/snunit.log
- PHP unit: Don't work  https://tfb-status.techempower.com/unzip/results.2023-03-23-04-22-59-528.zip/results/20230317041717/php-unit/run/php-unit.log

**Comment by joanhey** [comment]
With unit v1.31.0 also fail in the GH Actions tests in the bench.
`FROM unit:1.31.0-php8.2`

https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/6079079817/job/16491119686#step:9:5995

**Comment by ac000** [comment]
What is failing exactly?

**Comment by ac000** [comment]
I mean the only errors I'm seeing for php-unit are a bunch of [socket connection errors
](https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/6079079817/job/16491119686#step:9:6415)

```
Verifying test db for php-unit caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /dbraw.php (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff2bb167670>: Failed to establish a new connection: [Errno 111] Connection refused'))
   FAIL for http://tfb-server:8080
     Server did not respond to request
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
--------------------------------------------------------------------------------
VERIFYING JSON
--------------------------------------------------------------------------------
Accessing URL http://tfb-server:8080/json.php: 
Verifying test json for php-unit caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /json.php (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff2bb166bc0>: Failed to establish a new connection: [Errno 111] Connection refused'))
   FAIL for http://tfb-server:8080
     Server did not respond to request
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
--------------------------------------------------------------------------------
VERIFYING QUERY
--------------------------------------------------------------------------------
Accessing URL http://tfb-server:8080/dbquery.php?queries=2: 
Verifying test query for php-unit caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /dbquery.php?queries=2 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff2bb54be80>: Failed to establish a new connection: [Errno 111] Connection refused'))
   FAIL for http://tfb-server:8080
     Server did not respond to request
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
--------------------------------------------------------------------------------
VERIFYING UPDATE
--------------------------------------------------------------------------------
Accessing URL http://tfb-server:8080/updateraw.php?queries=2: 
Verifying test update for php-unit caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /updateraw.php?queries=2 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff2bb533b80>: Failed to establish a new connection: [Errno 111] Connection refused'))
   FAIL for http://tfb-server:8080
     Server did not respond to request
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
--------------------------------------------------------------------------------
VERIFYING FORTUNE
--------------------------------------------------------------------------------
Accessing URL http://tfb-server:8080/fortune.php: 
Verifying test fortune for php-unit caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /fortune.php (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff2bb56e3e0>: Failed to establish a new connection: [Errno 111] Connection refused'))
   FAIL for http://tfb-server:8080
     Server did not respond to request
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
--------------------------------------------------------------------------------
VERIFYING PLAINTEXT
--------------------------------------------------------------------------------
Accessing URL http://tfb-server:8080/plaintext.php: 
Verifying test plaintext for php-unit caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /plaintext.php (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff2bb56ee00>: Failed to establish a new connection: [Errno 111] Connection refused'))
   FAIL for http://tfb-server:8080
     Server did not respond to request
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
```

What is `tfb-server`? Is that what's running Unit? Could it be a file descriptor exhaustion issue?

**Comment by joanhey** [comment]
The last one is only a GH actions test:
https://github.com/TechEmpower/FrameworkBenchmarks/actions/runs/6079079817/job/16491119686#step:9:5995

I think the problem is that Unit needs too much time to initialize all the forks(processes|workers).

**Comment by joanhey** [comment]
Before Unit was working without problem in the benchmark. 
Later it start to fail in the benchmark, but now it also fail in the GitHub action test.

**Comment by ac000** [comment]
> I think the problem is that Unit needs too much time to initialize all the forks(processes|workers).

Are you saying that Unit itself is failing somehow due to trying to start X amount of processes?

What is X in your case?

Do you have a simple reproducer config?

**Comment by joanhey** [comment]
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/PHP/php/deploy/nginx-unit.json

It's a benchmark that use 3 xeon servers, each with 24 cores.

We test the same PHP code with different runtimes and servers.
The only failing is Unit, in the last versions of Unit.

You can download and run the tests locally. Before pass the GH action tests and fail only in the bench.
But now fail in both.

**Comment by ac000** [comment]
I'm just trying to understand exactly what it is that's failing...

Let me try it another way.

1) How many PHP application processes are you trying to start?

2) Do they all start correctly?

**Comment by joanhey** [comment]
I send it the link, but I pass it here:

```json
{
    "listeners": {
        "*:8080": {
            "pass": "applications/benchmark"
        }
    },

    "applications": {
        "benchmark": {
            "type": "php",
            "processes": 84,
            "user": "www-data",
            "group": "www-data",
            "root": "/php/",
            "options": {
                "file": "/php/deploy/conf/php.ini"
            },
            "limits": {
                "requests": 10000000
            }
        }
    }
}
```

**Comment by joanhey** [comment]
![image](https://github.com/nginx/unit/assets/249085/c8378797-daed-4f97-ae2a-d4a77a535338)

**Comment by joanhey** [comment]
You can clone the repo.
And later try it locally, because now also fail with the tests.

Only run in the root repo:  `./tfb --mode verify --test php-unit`

**Comment by ac000** [comment]
OK, so the issue started with 1.29.0...

If it was after that then I would maybe point to b0e2d9d0 ("Isolation: Switch to fork(2) & unshare(2) on Linux."), but it started failing before that landed.

Looking at the changes between 1.28.0 and 1.29.0, there's nothing that really stands out here.

There were a couple of PHP module updates, 48b6a7b3 ("PHP: Fixed php_module_startup() call for PHP 8.2.") and a0327445 ("PHP: allowed to specify URLs without a trailing '/'.")

The only obvious process related changes was the introduction of per application cgroup support, but you're not even using that so it should basically have no impact.

The thing I find confusing about the logs is it always looks like unit is shutdown before the benchmark even runs.

Does it work if you reduce the number of processes?

As a quick test, I was able to start 84 PHP application processes in < 1 second on a 2 vCPU VM.

```
2023/09/14 17:49:17 [info] 882#882 "php" prototype started
2023/09/14 17:49:17 [info] 883#883 "php" application started
2023/09/14 17:49:17 [info] 884#884 "php" application started
2023/09/14 17:49:17 [info] 885#885 "php" application started
2023/09/14 17:49:17 [info] 886#886 "php" application started
...
2023/09/14 17:49:17 [info] 964#964 "php" application started
2023/09/14 17:49:17 [info] 965#965 "php" application started
2023/09/14 17:49:17 [info] 966#966 "php" application started
```

**Comment by joanhey** [comment]
I don't remember the exact Unit version than start to fail.
But was just when I started this issue, so the version released before Nov 29, 2022.
![image](https://github.com/nginx/unit/assets/249085/6d9fdeb9-0146-4719-ac45-9277e86adb63)

I'll try if fail with less processes.

**Comment by joanhey** [comment]
Today I have some free time, and I try to pinpoint this issue.

The problem is that it needs too much time to start "Unit", and it isn't ready when start the verification test or any other CI command.

### Actual unit docker build process

1. Build the container
2. Finish the build
4. Run` ENTRYPOINT `( start unit | /docker-entrypoint.d/{*.sh, *.js, *.pem, *.json} | kill unit )
5. Run `CMD `(start unit)

The `ENTRYPOINT` and `CMD` are run after the build and need a lot of time, with **~70 processes**  it isn't fast enough to run the next CI command. The actual servers in the bench have a CPU with 56 threads.

### No `ENTRYPOINT` and config with `RUN`
If we bypass the `ENTRYPOINT` and config `unit` with `RUN` and tested with **168 processes** is really fast, and the next CI command run without problems.

```
RUN unitd && \
    curl -X PUT --data-binary @/php/deploy/nginx-unit.json --unix-socket \
        /var/run/control.unit.sock http://localhost/config

ENTRYPOINT [ ]

CMD ["unitd", "--no-daemon", "--control", "unix:/var/run/control.unit.sock"]
```
Perhaps I need to `kill unit` in the `RUN`.


**Welcome any feedback with another solution.**

**Comment by ac000** [comment]
> If we bypass the ENTRYPOINT and config unit with RUN and tested with 168 processes is really fast, and the next CI command run without problems.

Yes, that matches my previous comment where you can see starting Unit with a load load of PHP processes took < 1 second... so I suspect the problem is not Unit itself but around the machinery to start it...

I don't know what `ENTRYPOINT` is doing but perhaps you can instrument things to see where the time going?

**Comment by joanhey** [comment]
The `ENTRYPOINT`
https://github.com/nginx/unit/blob/master/pkg/docker/docker-entrypoint.sh

The partial log for`ENTRYPOINT` execution ( start unit | /docker-entrypoint.d/{*.sh, *.js, *.pem, *.json} | kill unit ) + `CMD` (start unit again) execution: 
https://tfb-status.techempower.com/unzip/results.2024-05-07-03-20-53-233.zip/results/20240501071818/php-unit/run/php-unit.log

We can also run the tests in local, and fail because use too much time with ~+70 processes.

**Comment by lcrilly** [comment]
See previous discussion: https://github.com/nginx/unit/discussions/1114

The entrypoint script works well for simple configurations/applications but does hurt container startup time.

The reason for the `start-config-kill-restart` process is because the Control API is the means of configuring Unit and so we need a running instance to actually do the configuration.

For my own Unit-based Docker images I take a similar approach with `RUN`. The entrypoint script will notice that a configuration is already set and will not overwrite it. So you don't need to set it empty.

```Dockerfile
COPY unitconf.json init.json
RUN nohup /bin/sh -c "unitd --no-daemon --pid init.pid --log /dev/stdout --control unix:init.unit.sock &" && \
    curl -fsX PUT -d@init.json --unix-socket init.unit.sock _/config && \
    rm init.*
```

I have a proposal for `unitd` to start in a non-daemonizing, "configuration mode" and read configuration from stdin. Which would also remove the dependency on having curl(1) in the container. I'll write an issue about it.

**Comment by lolgab** [comment]
@joanhey A workaround you can try is to have `max` and `spare` processes, so the first time Unit doesn't need to start 70 processes just to configure itself, it can start with `1` process and then scale to `84` when the real test begins:

```json
"processes": {
  "max": 84,
  "spare": 1
}
```

What do you think?

**Comment by joanhey** [comment]
Yes, I tried that without a solution.
They have a problem, and we can help them. But they need to find and fix the problems.

In local I moved the config to the entrypoint, and it's worst.
For me than they test the server with 80 concurrent requests.

I tried locally to update to PHP8.4, and it is NOT possible, fail completely.

I tried to help, but I send issues for the last 2 years, without any solution or they don't want see the problem.

**Comment by joanhey** [comment]
Please anybody from Unit core developers, can update Unit to PHP/8.4 ??

**Comment by joanhey** [comment]
Before Unit was faster than Nginx server.
But every run it's worst, less performance or not init at all.

**Comment by lolgab** [comment]
@joanhey Wasn't php updated in https://github.com/nginx/unit/issues/1491 ?

**Comment by joanhey** [comment]
Yes, but fail with each try to update the benchmark.

I tested 20 combinations, using the entrypoint, less processes, ... and never work.

If anybody from the core want to try, here it's my last dockerfile try:

```dockerfile
FROM unit:php8.4

RUN docker-php-ext-install pdo_mysql opcache > /dev/null

WORKDIR /php
COPY --link . .

RUN if [ $(nproc) = 2 ]; then sed -i "s|\"spare\": 168,|\"spare\": 64,|g" /php/deploy/nginx-unit.json ; fi;

#RUN more /php/deploy/nginx-unit.json


COPY deploy/nginx-unit.json /docker-entrypoint.d/

#ENTRYPOINT [ ]

EXPOSE 8080


#CMD ["unitd", "--no-daemon", "--control", "unix:/var/run/control.unit.sock"]
```

The last time with PHP8.3 I needed to configure first and bypass the entrypoint.

**Comment by ac000** [comment]
Hi @joanhey 

> I tried to help, but I send issues for the last 2 years, without any solution or they don't want see the problem.

It's not that we don't *want* to see the problem... but the problem seems to be related to the whole cycle of starting unit/configuring it/stopping it/starting it again...

That is certainly less than ideal, but can be easily avoided (by for example, and this is just one possibility, using a pre-populated state dir, yes, people will say that's not *supported* or whatever, but in reality it'll work most of the time).

Looking back at some of the recent comments

In https://github.com/nginx/unit/issues/794#issuecomment-1719829378 I showed that Unit has no problem starting many php application processes in < 1 second.

Then in https://github.com/nginx/unit/issues/794#issuecomment-2118961911 you yourself say

> If we bypass the ENTRYPOINT and config unit with RUN and tested with 168 processes is really fast, and the next CI command run without problems.

So, don't use the entrypoint.sh thing which is known to have issues, seems to be the solution...
